In [13]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## reading clean Data


In [14]:
data = pd.read_csv('data/raw/data.csv',index_col=0)
data.head()

uniqueid datavaluetypeid  populationcount  ACCESS2  ARTHRITIS  \
0              0107000       AgeAdjPrv         212237.0    0.198      0.310   
1              0107000          CrdPrv         212237.0    0.196      0.309   
2  0107000-01073000100          CrdPrv           3042.0    0.239      0.325   
3  0107000-01073000300          CrdPrv           2735.0    0.288      0.313   
4  0107000-01073000400          CrdPrv           3338.0    0.261      0.346   

   BINGE  BPHIGH  BPMED  CANCER  CASTHMA  ...  KIDNEY    LPA  MAMMOUSE  MHLTH  \
0  0.112   0.441  0.701   0.057    0.115  ...   0.035  0.373     0.797  0.156   
1  0.113   0.438  0.813   0.057    0.115  ...   0.036  0.370     0.768  0.156   
2  0.101   0.462  0.800   0.052    0.127  ...   0.036  0.423     0.742  0.188   
3  0.108   0.456  0.803   0.046    0.118  ...   0.038  0.437     0.740  0.185   
4  0.095   0.501  0.820   0.052    0.130  ...   0.040  0.449     0.736  0.191   

   OBESITY  PAPTEST  PHLTH  SLEEP  STROKE  TEETHLOST  
0    0.427    0.801  0.164  0.469   0.052      0.259  
1    0.419    0.786  0.164  0.465   0.052      0.261  
2    0.467    0.760  0.196  0.504   0.055      0.302  
3    0.470    0.732  0.202  0.494   0.057      0.333  
4    0.488    0.728  0.211  0.525   0.065      0.360  

[5 rows x 31 columns]

In [15]:
locations = pd.read_csv('data/raw/locations.csv',index_col=0)
locations.head()

uniqueid stateabbr    cityname geographiclevel  \
6325               0107000        AL  Birmingham            City   
40098  0107000-01073000100        AL  Birmingham    Census Tract   
2125   0107000-01073000300        AL  Birmingham    Census Tract   
20523  0107000-01073000400        AL  Birmingham    Census Tract   
1727   0107000-01073000500        AL  Birmingham    Census Tract   

          tractfips  cityfips statedesc   latitude  longitude  
6325            NaN  107000.0   Alabama -86.798817  33.527566  
40098  1.073000e+09  107000.0   Alabama -86.722832  33.579433  
2125   1.073000e+09  107000.0   Alabama -86.752434  33.542821  
20523  1.073000e+09  107000.0   Alabama -86.764047  33.563245  
1727   1.073000e+09  107000.0   Alabama -86.774913  33.544240

In [16]:
locations.columns

Index(['uniqueid', 'stateabbr', 'cityname', 'geographiclevel', 'tractfips',
       'cityfips', 'statedesc', 'latitude', 'longitude'],
      dtype='object')

In [17]:
locations.cityname = locations.cityname.fillna('United States') ##INSTEAD OF NAN, fill CITY OF UNITED STATES FOR AGG FUNCTIONS

In [18]:
dataTemp = data.set_index('uniqueid').join(locations.set_index('uniqueid'))
dataTemp['geographiclevel'].value_counts()
#dataTemp[dataTemp['datavaluetypeid']=='AgeAdjPrv']

Census Tract    27198
City             1000
US                  2
Name: geographiclevel, dtype: int64

In [19]:
dataTemp.columns

Index(['datavaluetypeid', 'populationcount', 'ACCESS2', 'ARTHRITIS', 'BINGE',
       'BPHIGH', 'BPMED', 'CANCER', 'CASTHMA', 'CHD', 'CHECKUP', 'CHOLSCREEN',
       'COLON_SCREEN', 'COPD', 'COREM', 'COREW', 'CSMOKING', 'DENTAL',
       'DIABETES', 'HIGHCHOL', 'KIDNEY', 'LPA', 'MAMMOUSE', 'MHLTH', 'OBESITY',
       'PAPTEST', 'PHLTH', 'SLEEP', 'STROKE', 'TEETHLOST', 'stateabbr',
       'cityname', 'geographiclevel', 'tractfips', 'cityfips', 'statedesc',
       'latitude', 'longitude'],
      dtype='object')

# missing Values check


In [20]:
#data = data.join(locations[['cityname','geographiclevel','uniqueid']].set_index('uniqueid'),on='uniqueid',how='left',)
data.describe().sort_values('count',axis=1)

COREW         COREM     TEETHLOST      MAMMOUSE  COLON_SCREEN  \
count  28091.000000  28097.000000  28140.000000  28163.000000  28181.000000   
mean       0.287004      0.299446      0.162323      0.749657      0.593809   
std        0.072547      0.064604      0.084670      0.050455      0.092823   
min        0.096000      0.131000      0.029000      0.414000      0.234000   
25%        0.231000      0.250000      0.096000      0.718000      0.529000   
50%        0.287000      0.299000      0.144000      0.756000      0.607000   
75%        0.339000      0.346000      0.212000      0.788000      0.666000   
max        0.538000      0.522000      0.550000      0.889000      0.815000   

            PAPTEST       ACCESS2         BINGE         SLEEP         PHLTH  \
count  28193.000000  28199.000000  28200.000000  28200.000000  28200.000000   
mean       0.789161      0.174237      0.167900      0.370524      0.129222   
std        0.063533      0.098298      0.038247      0.063665      0.043165   
min        0.376000      0.024000      0.024000      0.185000      0.031000   
25%        0.751000      0.099000      0.144000      0.325000      0.096000   
50%        0.799000      0.151000      0.165000      0.363000      0.122000   
75%        0.838000      0.227000      0.189000      0.411000      0.157000   
max        0.915000      0.641000      0.432000      0.598000      0.377000   

       ...          COPD    CHOLSCREEN       CHECKUP           CHD  \
count  ...  28200.000000  28200.000000  28200.000000  28200.000000   
mean   ...      0.061554      0.726560      0.683670      0.056381   
std    ...      0.024407      0.083061      0.061276      0.020961   
min    ...      0.009000      0.217000      0.454000      0.002000   
25%    ...      0.044000      0.678000      0.639750      0.043000   
50%    ...      0.057000      0.734000      0.681000      0.054000   
75%    ...      0.074000      0.785000      0.727000      0.067000   
max    ...      0.250000      0.955000      0.894000      0.346000   

            CASTHMA        CANCER         BPMED        BPHIGH        DENTAL  \
count  28200.000000  28200.000000  28200.000000  28200.000000  28200.000000   
mean       0.094804      0.055842      0.719670      0.304917      0.598970   
std        0.017219      0.017869      0.078521      0.082096      0.130059   
min        0.053000      0.007000      0.115000      0.049000      0.189000   
25%        0.082000      0.044000      0.680000      0.254000      0.502000   
50%        0.091000      0.054000      0.730000      0.293000      0.614000   
75%        0.104000      0.065000      0.772000      0.344000      0.704000   
max        0.193000      0.231000      0.932000      0.710000      0.871000   

       populationcount  
count     2.820000e+04  
mean      3.285559e+04  
std       2.601662e+06  
min       5.100000e+01  
25%       2.500000e+03  
50%       3.694500e+03  
75%       5.080000e+03  
max       3.087455e+08  

[8 rows x 29 columns]

7 columns missing 

In [21]:
# pull dataframe of missing data, join city and geography type, and get a count by geography type.
missingData = data[data.isnull().values].join(locations[['uniqueid','cityname','geographiclevel']].set_index('uniqueid'),on='uniqueid',how='left')

#value counts of missing data by geographic level
print(missingData['geographiclevel'].value_counts(),end='\n\n')

#value counts of all data by geographic level
print(data.join(locations[['uniqueid','cityname','geographiclevel']].set_index('uniqueid'),on='uniqueid',how='left')['geographiclevel'].value_counts())

Census Tract    336
Name: geographiclevel, dtype: int64

Census Tract    27198
City             1000
US                  2
Name: geographiclevel, dtype: int64


#### As we can see there are only 336 census tracts of missing data out of 27198 (1%). No rolled up cities are missing data as well as the rolled up united states.
#### Now I'll build a dataframe that includes only missing data grouped by city, showing the mean population of the census tracts with missing data, the count of census tracts in each city missing data, the count of total census tracts from those cities, and the % of census tracts that are missing from each city

In [22]:
missingDataPopMean = missingData.groupby('cityname')['populationcount'].mean()
missingDataCensusCount = missingData.groupby('cityname')['uniqueid'].count()
missingDataJoined = pd.concat([missingDataPopMean,missingDataCensusCount],axis=1,keys=['AvgPop','MissingCensusCount'])
fullDataCensusCount = data.join(locations[['uniqueid','cityname','geographiclevel']].set_index('uniqueid'),on='uniqueid',how='left').groupby('cityname')['uniqueid'].count()
missingDataCities = missingDataJoined.join(fullDataCensusCount,how='left')
missingDataCities.columns = ['Avg Population','Num Census Tracts Missing','CensusTract Count']
missingDataCities['perMissing'] = missingDataCities.iloc[:,1]/missingDataCities.iloc[:,2]
missingDataCities.sort_values('perMissing',ascending=False,inplace=True)
print('TOP AND BOTTOM 5 Cities Missing Percentage Of Censuses')
pd.concat([missingDataCities.head(10)])

TOP AND BOTTOM 5 Cities Missing Percentage Of Censuses


Avg Population  Num Census Tracts Missing  CensusTract Count  \
cityname                                                                        
Santa Maria          147.000000                          6                 20   
Fayetteville        1709.733333                         15                 69   
Rock Hill           1762.000000                          5                 24   
Gulfport              85.000000                          5                 24   
Missouri City         56.750000                          4                 20   
Abilene             1700.428571                          7                 39   
Boca Raton          2224.000000                          5                 29   
Miami Beach           59.000000                          5                 31   
Portsmouth            78.000000                          5                 33   
College Station     6964.250000                          4                 28   

                 perMissing  
cityname                     
Santa Maria        0.300000  
Fayetteville       0.217391  
Rock Hill          0.208333  
Gulfport           0.208333  
Missouri City      0.200000  
Abilene            0.179487  
Boca Raton         0.172414  
Miami Beach        0.161290  
Portsmouth         0.151515  
College Station    0.142857

### Should we drop any census tracts from the data? IE is the  census missing most of it's data across all variables?

In [23]:
temp = data.drop('datavaluetypeid',axis=1).set_index('uniqueid')
temp.head()

populationcount  ACCESS2  ARTHRITIS  BINGE  BPHIGH  \
uniqueid                                                                  
0107000                     212237.0    0.198      0.310  0.112   0.441   
0107000                     212237.0    0.196      0.309  0.113   0.438   
0107000-01073000100           3042.0    0.239      0.325  0.101   0.462   
0107000-01073000300           2735.0    0.288      0.313  0.108   0.456   
0107000-01073000400           3338.0    0.261      0.346  0.095   0.501   

                     BPMED  CANCER  CASTHMA    CHD  CHECKUP  ...  KIDNEY  \
uniqueid                                                     ...           
0107000              0.701   0.057    0.115  0.076    0.768  ...   0.035   
0107000              0.813   0.057    0.115  0.077    0.767  ...   0.036   
0107000-01073000100  0.800   0.052    0.127  0.080    0.763  ...   0.036   
0107000-01073000300  0.803   0.046    0.118  0.082    0.738  ...   0.038   
0107000-01073000400  0.820   0.052    0.130  0.088    0.775  ...   0.040   

                       LPA  MAMMOUSE  MHLTH  OBESITY  PAPTEST  PHLTH  SLEEP  \
uniqueid                                                                      
0107000              0.373     0.797  0.156    0.427    0.801  0.164  0.469   
0107000              0.370     0.768  0.156    0.419    0.786  0.164  0.465   
0107000-01073000100  0.423     0.742  0.188    0.467    0.760  0.196  0.504   
0107000-01073000300  0.437     0.740  0.185    0.470    0.732  0.202  0.494   
0107000-01073000400  0.449     0.736  0.191    0.488    0.728  0.211  0.525   

                     STROKE  TEETHLOST  
uniqueid                                
0107000               0.052      0.259  
0107000               0.052      0.261  
0107000-01073000100   0.055      0.302  
0107000-01073000300   0.057      0.333  
0107000-01073000400   0.065      0.360  

[5 rows x 29 columns]

In [24]:
# How many data points are any of these census's missing?
sm = data.drop('datavaluetypeid',axis=1).set_index('uniqueid').apply(lambda x: np.where(np.isnan(x),1,0),axis=1).sum(axis=1)
sm.name ='numMissing'
ct = data.drop('datavaluetypeid',axis=1).set_index('uniqueid').apply(lambda x: np.where(np.isnan(x),1,0),axis=1).count(axis=1)
div = sm.divide(ct)
div.name = '% Missing'
locations.set_index('uniqueid').loc[sm.divide(ct).sort_values(ascending=False).head(25).index,:].join(sm).sort_values('numMissing',ascending=False).join(div)

ValueError: No axis named 1 for object type <class 'pandas.core.series.Series'>

#### It looks like the most amount of variables any one census is missing is 21%. I'll keep all census tracts in.

## Clean Data by filling census tract NANs with data from mean values of the same city 

In [ ]:
cleanedData = data.join(locations[['uniqueid','cityname','stateabbr','geographiclevel']].set_index('uniqueid'),on='uniqueid',how='left').copy()
colsToClean = cleanedData.drop(['uniqueid','cityname','stateabbr','datavaluetypeid','populationcount','geographiclevel'],axis=1).columns
for col in colsToClean:
    cleanedData[col] = cleanedData.groupby('cityname')[col].transform(lambda x: x.fillna(np.mean(x)))

### try median for missing values

In [ ]:
# Check cleaned Data For Missing Values. 
cleanedData.describe().sort_values('count',axis=1)

### Are there any outliers in the data that need to be cleaned?

In [ ]:
fig, ax = plt.subplots(1,1,figsize=(20,12))
sns.boxplot(data=cleanedData.drop(['populationcount'],axis=1),ax=ax,orient='horizontal')
plt.show()

When looking at a box and whisker plot, there are many census tracts that fall outside of of the inter quartile ranges (outliers) but I don't believe they are erroneous data.

This could be cause by low sample size of certain census tracts or there are truly many census tracts that fall outside of the inter quartile ranges.

Now we'll look at the outliers at a city level to see if there's much of a difference

In [ ]:
tst = cleanedData[cleanedData['geographiclevel']=='Census Tract'].set_index('uniqueid').drop(['cityname','stateabbr','geographiclevel','datavaluetypeid','populationcount'],axis=1)
tst1 = tst.multiply(cleanedData[cleanedData['geographiclevel']=='Census Tract'].set_index('uniqueid')['populationcount'],axis=0)
tst2 = tst1.join(cleanedData[cleanedData['geographiclevel']=='Census Tract'].set_index('uniqueid')[['stateabbr','cityname']]).groupby(['stateabbr','cityname']).agg(np.sum)
tst3 = tst2.divide(cleanedData[cleanedData['geographiclevel']=='Census Tract'].groupby(['stateabbr','cityname'])['populationcount'].sum(),axis=0)

fig, ax = plt.subplots(1,1,figsize=(20,12))
_  = sns.boxplot(data=tst3,orient='horizontal')
plt.show()


It looks like most of the data points are somewhat of a normal distribution. I think this data is clean and ready for a deeper analysis. Time to save it for future Use!

In [ ]:
cleanedData.drop(['geographiclevel','stateabbr','cityname'],axis=1).to_csv('../data/cleansed/data.csv')
locations.to_csv('../data/cleansed/locations.csv')